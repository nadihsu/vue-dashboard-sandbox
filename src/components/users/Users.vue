<template lang="pug">
div
  .title {{ $t('M_USER_TITLE') }}

  el-button(type="primary", icon="Edit", @click="showCreateModal = true") {{ $t('ACTION.CREATED') }}

  SearchUser(:get-users="getUsers")
  UserList(:data="data", :edit-user="editUser", :delete-user="deleteUser")
  .pagination-wrapper
    el-pagination(
    background,
    layout="total, prev, pager, next, sizes",
    :page-size="Number(pagination.max_results)",
    :page-sizes="[20, 40, 100]",
    :total="pagination.total",
    @size-change="handleSizeChange",
    @current-change="handlePagedChange").

el-dialog(v-model="showCreateModal", :title="$t('M_CREATED_USER')", width="25%")
  template(#default)
    EditUserModal(
    :active="showCreateModal",
    :close-modal="onClose",
    :create-user="createUser",
    :get-users="getUsers")
</template>

<script lang="ts">
import { ref, onMounted, defineComponent } from 'vue';
import { request } from '@/utils';
import UserList from './UserList.vue';
import EditUserModal from './EditUserModal.vue';
import SearchUser from './SearchUser.vue';

export default defineComponent({
  components: {
    UserList,
    EditUserModal,
    SearchUser,
  },
  setup() {
    const showCreateModal = ref(false);
    const pager = ref(2);
    const data = ref([]);
    const pagination = ref({
      first_result: 0,
      max_results: 20,
      total: 0,
    });

    /**
     * 處理查詢條件
     *
     * @param dataData 查詢條件
     * @returns 過濾後查詢條件
     */
    function handleSubmitQuery(dataData:object) {
      const result:{ [key: string]: number | string } = {};

      Object.entries(dataData)
        .filter(([, val]) => val !== -1) // 選項不限
        .filter(([, val]) => (val !== '')) // 過濾沒有值的選項
        .forEach(([key, val]) => {
          result[key] = val;
        });

      return result;
    }

    /**
     * 取得使用者
     *
     * @param param params
     * @param param.isSearching 是否檢索
     * @param param.newPagination 新頁碼
     * @param param.searchData 查詢條件
     */
    async function getUsers({
      isSearching = false,
      newPagination = pagination.value,
      searchData = {},
    } = {}) {
      const qs = handleSubmitQuery(searchData);

      const payload = {
        ...qs,
        first_result: isSearching ? 0 : newPagination.first_result,
        max_results: newPagination.max_results,
      };

      const out = await request('GET', '/users', payload);

      if (out?.result === 'ok') {
        data.value = out.ret;
        pagination.value = out.pagination;
      }
    }

    /**
     * 新增使用者
     *
     * @param userData 新增的資料
     */
    async function createUser(userData:object) {
      const out = await request('POST', '/user', userData);

      if (out?.result === 'ok') {
        getUsers();
      }
    }

    /**
     * 編輯使用者
     *
     * @param userId 使用者代號
     * @param userData 變更使用者資料
     */
    async function editUser(userId:string, userData:object) {
      const out = await request('PUT', `/user/${userId}`, userData);

      if (out?.result === 'ok') {
        getUsers();
      }
    }

    /**
     * 刪除使用者
     *
     * @param userId 使用者代號
     */
    async function deleteUser(userId:string) {
      const out = await request('DELETE', `/user/${userId}`);

      if (out?.result === 'ok') {
        getUsers();
      }
    }

    /**
     * 切換頁碼
     *
     * @param pageNo 頁碼數
     */
    function handlePagedChange(pageNo:number) {
      const newPagination = {
        ...pagination.value,
        first_result: pagination.value.max_results * (pageNo - 1),
      };

      getUsers({ newPagination });
      pager.value += 1;
    }

    /**
     * 切換頁碼總數
     *
     * @param val 每頁頁碼總數
     */
    function handleSizeChange(val:number) {
      const newPagination = {
        ...pagination.value,
        max_results: val,
      };

      getUsers({ newPagination });
    }

    /**
     * 關閉新增彈窗
     */
    function onClose() {
      showCreateModal.value = false;
    }

    onMounted(() => {
      getUsers({ isSearching: true });
    });

    return {
      showCreateModal,
      pager,
      data,
      pagination,
      handleSubmitQuery,
      getUsers,
      createUser,
      editUser,
      deleteUser,
      handlePagedChange,
      handleSizeChange,
      onClose,
    };
  },
});
</script>

<style scoped>
  .pagination-wrapper {
    margin-top: 1rem;
  }
</style>
