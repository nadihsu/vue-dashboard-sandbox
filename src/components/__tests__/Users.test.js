import { render } from 'Tests/unit/TestUtils';
import Users from '../users/Users.vue';

describe('Users Page', () => {
  it('基本', () => {
    const setup = render(Users, {
      globalConfig: {
        stubs: ['UserList', 'EditUserModal', 'SearchUser'],
      },
    });

    expect(setup.html()).toMatchSnapshot();
  });
});
