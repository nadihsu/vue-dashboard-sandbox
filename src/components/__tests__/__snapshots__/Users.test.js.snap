// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Users Page > 基本 1`] = `
<div>
  <div class="title">使用者名單</div>
  <button aria-disabled="false" class="el-button el-button--primary" type="button">
    <i class="el-icon">
      <edit></edit>
    </i>
    <span class>新增</span>
  </button>
  <search-user-stub getusers="[Function]"></search-user-stub>
  <user-list-stub data deleteuser="[Function]" edituser="[Function]"></user-list-stub>
  <div class="pagination-wrapper">
    <div class="el-pagination is-background">
      <span class="el-pagination__total is-first" disabled="false">Total 0</span>
      <button aria-disabled="true" aria-label="Go to previous page" class="btn-prev" disabled type="button">
        <i class="el-icon">
          <svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
            <path d="M609.408 149.376 277.76 489.6a32 32 0 0 0 0 44.672l331.648 340.352a29.12 29.12 0 0 0 41.728 0 30.592 30.592 0 0 0 0-42.752L339.264 511.936l311.872-319.872a30.592 30.592 0 0 0 0-42.688 29.12 29.12 0 0 0-41.728 0z" fill="currentColor" />
          </svg>
        </i>
      </button>
      <ul class="el-pager">
        <li aria-current="true" aria-label="page 1" class="is-active number" tabindex="0"> 1 </li>
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
      </ul>
      <button aria-disabled="true" aria-label="Go to next page" class="btn-next" disabled type="button">
        <i class="el-icon">
          <svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
            <path d="M340.864 149.312a30.592 30.592 0 0 0 0 42.752L652.736 512 340.864 831.872a30.592 30.592 0 0 0 0 42.752 29.12 29.12 0 0 0 41.728 0L714.24 534.336a32 32 0 0 0 0-44.672L382.592 149.376a29.12 29.12 0 0 0-41.728 0z" fill="currentColor" />
          </svg>
        </i>
      </button>
      <span class="el-pagination__sizes is-last">
        <div class="el-select el-select--default">
          <div class="select-trigger el-tooltip__trigger el-tooltip__trigger">
            <!--v-if-->
            <!-- fix: https://github.com/element-plus/element-plus/issues/11415 -->
            <!--v-if-->
            <div class="el-input el-input--default el-input--suffix">
              <!-- input -->
              <!-- prepend slot -->
              <!--v-if-->
              <div class="el-input__wrapper">
                <!-- prefix slot -->
                <!--v-if-->
                <input autocomplete="off" class="el-input__inner" placeholder="Select" readonly tabindex="0" type="text"><!-- suffix slot -->
                <span class="el-input__suffix">
                  <span class="el-input__suffix-inner">
                    <i class="el-icon el-select__caret el-select__icon">
                      <svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
                        <path d="M831.872 340.864 512 652.672 192.128 340.864a30.592 30.592 0 0 0-42.752 0 29.12 29.12 0 0 0 0 41.6L489.664 714.24a32 32 0 0 0 44.672 0l340.288-331.712a29.12 29.12 0 0 0 0-41.728 30.592 30.592 0 0 0-42.752 0z" fill="currentColor" />
                      </svg>
                    </i>
                    <!--v-if-->
                    <!--v-if-->
                    <!--v-if-->
                    <!--v-if-->
                    <!--v-if-->
                    <!--v-if-->
                  </span>
                </span>
              </div><!-- append slot -->
              <!--v-if-->
            </div>
          </div>
          <!--teleport start-->
          <!--teleport end-->
        </div>
      </span>
    </div>
  </div>
</div>
<!--teleport start-->
<transition-stub appear="false" css="true" name="dialog-fade" persisted="true">
  <div class="el-overlay" style="z-index: 2002; display: none;">
    <div aria-describedby aria-label="新增使用者" aria-modal="true" class="el-overlay-dialog" role="dialog"></div>
  </div>
</transition-stub>
<!--teleport end-->
`;
